src/Sulu/Component/HttpCache/EventSubscriber/InvalidationSubscriber.php

```diff
@@ -57,25 +93,122 @@ public function __construct(
+            return;
+        }
+
+        // invalidate urls of the locale of the document
+        $locale = $this->documentInspector->getLocale($document);
+        foreach ($this->getLocaleUrls($document, $locale) as $url) {
+            $this->pathHandler->invalidatePath($url);
+        }
+    }
+
+    /**
+     * Invalidates all urls in all locales of the document when a document gets removed.
+     * This method is executed before the actual removing of the document because the document must still
+     * exist to gather the urls of the document.
+     *
+     * @param RemoveEvent $event
+     */
+    public function invalidateDocumentBeforeRemoving(RemoveEvent $event)
     {
         $document = $event->getDocument();
 
-        if (!$document instanceof StructureBehavior) {
+        if (!$document instanceof ResourceSegmentBehavior) {
             return;
         }
 
-        $structureBridge = $this->structureManager->wrapStructure(
-            $this->documentInspector->getMetadata($document)->getAlias(),
-            $this->documentInspector->getStructureMetadata($document)
-        );
+        // invalidate urls of all locales of the document
+        foreach ($this->documentInspector->getPublishedLocales($document) as $locale) {
+            foreach ($this->getLocaleUrls($document, $locale) as $url) {
+                $this->pathHandler->invalidatePath($url);
+            }
+        }
+    }
+
+    /**
+     * Returns all urls of the given locale which are associated with the given document.
+     * The returned array contains all current urls and all history urls.
+     * The returned urls can contain placeholders (eg {host}).
+     *
+     * @param ResourceSegmentBehavior $document
+     * @param $locale
+     *
+     * @return array Urls of the given locale which are associated with the given document
+     */
+    private function getLocaleUrls(ResourceSegmentBehavior $document, $locale)
+    {
+        $documentUuid = ($document instanceof UuidBehavior) ? $document->getUuid() : null;
+        $webspace = ($document instanceof WebspaceBehavior) ? $document->getWebspaceName() : null;
+
+        if (!$locale || !$documentUuid) {
+            return [];
+        }
+
+        // get current resource-locator and history resource-locators
+        $resourceLocators = [];
+        $resourceLocators[] = $this->rlpStrategy->loadByContentUuid($documentUuid, $webspace, $locale);
+
+        $historyResourceLocators = $this->rlpStrategy->loadHistoryByContentUuid($documentUuid, $webspace, $locale);
+        foreach ($historyResourceLocators as $historyResourceLocator) {
+            $resourceLocators[] = $historyResourceLocator->getResourceLocator();
+        }
+
+        // get urls for resource-locators
+        $urls = [];
+        foreach ($resourceLocators as $resourceLocator) {
+            $rlpUrls = $this->webspaceManager->findUrlsByResourceLocator(
```

Comment:
Don't use `rlp`. Actually you could just call this method in the `array_merge` call, this would make the variable obsolete at all.